ext/uri/php_lexbor.c

```diff
@@ -0,0 +1,604 @@
+/*
+   +----------------------------------------------------------------------+
+   | Copyright (c) The PHP Group                                          |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | https://www.php.net/license/3_01.txt                                 |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Authors: Máté Kocsis <kocsismate@php.net>                            |
+   +----------------------------------------------------------------------+
+*/
+
+#include "php.h"
+#include "php_lexbor.h"
+#include "php_uri_common.h"
+#include "Zend/zend_enum.h"
+#include "Zend/zend_smart_str.h"
+#ifdef HAVE_ARPA_INET_H
+#include <arpa/inet.h>
+#endif
+
+static zend_result lexbor_init_parser(void);
+static void *lexbor_parse_uri(const zend_string *uri_str, const void *base_url, zval *errors);
+static void lexbor_create_invalid_uri_exception(zval *exception_zv, zval *errors);
+static void *lexbor_clone_uri(void *uri);
+static zend_string *lexbor_uri_to_string(void *uri, uri_recomposition_mode_t recomposition_mode, bool exclude_fragment);
+static void lexbor_free_uri(void *uri);
+static zend_result lexbor_destroy_parser(void);
+
+HashTable lexbor_property_handlers;
+
+lxb_url_parser_t *lexbor_parser;
+int lexbor_urls;
+
+const uri_handler_t lexbor_uri_handler = {
+	URI_PARSER_WHATWG,
+	lexbor_init_parser,
+	lexbor_parse_uri,
+	lexbor_create_invalid_uri_exception,
+	lexbor_clone_uri,
+	lexbor_uri_to_string,
+	lexbor_free_uri,
+	lexbor_destroy_parser,
+	&lexbor_property_handlers
+};
+
+#define ZVAL_TO_LEXBOR_STR(value, str) do { \
+	if (Z_TYPE_P(value) == IS_STRING && Z_STRLEN_P(value) > 0) { \
+		lexbor_str_init_append(&str, lexbor_parser->mraw, (const lxb_char_t *) Z_STRVAL_P(value), Z_STRLEN_P(value)); \
+	} else if (Z_TYPE_P(value) == IS_LONG && Z_LVAL_P(value) != 0) { \
+	    ZVAL_STR(value, zend_long_to_str(Z_LVAL_P(value))); \
+		lexbor_str_init_append(&str, lexbor_parser->mraw, (const lxb_char_t *) Z_STRVAL_P(value), Z_STRLEN_P(value)); \
+	} else { \
+		lexbor_str_init(&str, lexbor_parser->mraw, 0); \
+	} \
+} while (0)
+
+#define CHECK_WRITE_RESULT(status, lexbor_uri, str, errors) do { \
+	if (status != LXB_STATUS_OK) { \
+		fill_errors(errors); \
+		return FAILURE; \
+	} \
+	return SUCCESS; \
+} while (0)
```

Comment:
I feel that this macro is not much simpler than simply using copy and paste of the definition. To me it makes the code much harder to follow, though, since it is not clear that it unconditionally returns and since it also has unused parameters.